src/comgr.cpp

```diff
@@ -0,0 +1,490 @@
+        return sz;
+    }
+
+    public:
+    std::size_t GetBytes(std::vector<char>& bytes) const
+    {
+        std::size_t sz = GetSize();
+        bytes.resize(sz);
+        ECI_THROW(amd_comgr_get_data(handle, &sz, &bytes[0]), sz);
+        return sz;
+    }
+    std::string GetString() const
+    {
+        std::vector<char> bytes;
+        const auto sz = GetBytes(bytes);
+        return {&bytes[0], sz};
+    }
+};
+
+class Dataset : ComgrOwner
+{
+    amd_comgr_data_set_t handle = {0};
+
+    public:
+    Dataset() { EC_THROW(amd_comgr_create_data_set(&handle)); }
+    ~Dataset() { EC(amd_comgr_destroy_data_set(handle)); }
+    auto operator()() const { return handle; }
+    void AddData(const Data& d) const { EC_THROW(amd_comgr_data_set_add(handle, d())); }
+    void AddData(const std::string& name,
+                 const std::string& content,
+                 const amd_comgr_data_kind_t type) const
+    {
+        const Data d(type);
+        MIOPEN_LOG_I2(name << ' ' << content.size() << " bytes");
+        d.SetName(name);
+        d.SetBytes(content);
+        AddData(d);
+        const auto show_first = miopen::Value(MIOPEN_DEBUG_COMGR_LOG_SOURCE_TEXT{}, 0);
+        if(show_first > 0 && miopen::IsLogging(miopen::LoggingLevel::Info) &&
+           type == AMD_COMGR_DATA_KIND_SOURCE)
+        {
+            const auto text_length = (content.size() > show_first) ? show_first : content.size();
+            const std::string text(content, 0, text_length);
+            MIOPEN_LOG_I(text);
+        }
+    }
+    size_t GetDataCount(const amd_comgr_data_kind_t kind) const
+    {
+        std::size_t count = 0;
+        ECI_THROW(amd_comgr_action_data_count(handle, kind, &count), kind);
+        return count;
+    }
+    Data GetData(const amd_comgr_data_kind_t kind, const std::size_t index) const
+    {
+        amd_comgr_data_t d;
+        ECI_THROW(amd_comgr_action_data_get_data(handle, kind, index, &d), kind);
+        return {d};
+    }
+};
+
+class ActionInfo : ComgrOwner
+{
+    amd_comgr_action_info_t handle = {0};
+
+    public:
+    ActionInfo() { EC_THROW(amd_comgr_create_action_info(&handle)); }
+    ~ActionInfo() { EC(amd_comgr_destroy_action_info(handle)); }
+    auto operator()() const { return handle; }
+    void SetLanguage(const amd_comgr_language_t language) const
+    {
+        ECI_THROW(amd_comgr_action_info_set_language(handle, language), language);
+    }
+    void SetIsaName(const std::string& isa) const
+    {
+        ECI_THROW_MSG(amd_comgr_action_info_set_isa_name(handle, isa.c_str()), isa, isa);
+    }
+    void SetLogging(const bool state) const
+    {
+        ECI_THROW(amd_comgr_action_info_set_logging(handle, state), state);
+    }
+    void SetOptionList(const std::vector<std::string>& options) const
+    {
+        std::vector<const char*> vp;
+        vp.reserve(options.size());
+        for(auto& opt : options) // cppcheck-suppress useStlAlgorithm
+            vp.push_back(opt.c_str());
```

Comment:
Use `std::transform` here.

Reply:
Oh no. Tried this and got horrible construct, almost impossible to read.

Reply:
... and IIRC after that cppcheck or tidy begun to complain about something else so I decided to give up and shut it.

Reply:
You should be able to write:

```cpp
std::transform(options.begin(), options.end(), std::back_inserter(vp), [&](auto& opt) {
    return opt.c_str();
});
```

That is not horrible to read, and it is actually much clearer to what the loop is doing since it has a name associated with it.

The `cppcheck-suppress` should be used for false positives and not because we dont like the check.

Reply:
> The cppcheck-suppress should be used for false positives and not because we dont like the check.

This is value judgment that does not cover all possibilities. For example checks can be useful in general but too stupid in corner cases.

I will try your snippet of course, thanks.

Reply:
Works fine. Done.